pysph/base/opencl.py

```diff
@@ -35,6 +42,36 @@ def set_queue(q):
     _queue = q
 
 
+def profile(name, event):
+    global _profile_info
+    event.wait()
+    time = (event.profile.end - event.profile.start) * 1e-9
+    _profile_info[name] += time
+
+
+def print_profile():
+    global _profile_info
+    _profile_info = sorted(_profile_info.iteritems(), key=itemgetter(1),
+                           reverse=True)
+    if len(_profile_info) == 0:
+        print("No profile information available")
+        return
+    print("{:<30} {:<30}".format('Kernel', 'Time'))
+    tot_time = 0
+    for kernel, time in _profile_info:
+        print("{:<30} {:<30}".format(kernel, time))
+        tot_time += time
+    print("Total profiled time: %g secs" % tot_time)
+
+
+def profile_kernel(kernel, name):
+    def _profile_knl(*args):
```

Comment:
You can put the if outside so if profiling is on, it just returns the original function and if not, it calls profile after the kernel invocation.  Also, you could return the event from the wrapped function just in case someone wants to use it to wait or something.